text/0074/0074-entity-slicing.md

```diff
@@ -0,0 +1,566 @@
+  resource
+)
+when
+{
+  resource.readers.contains(principal)
+};
+
+// allow documents to be read and edited by owners
+permit(
+  principal,
+  action in [Action::"Read", Action::"Edit"],
+  resource
+)
+when
+{
+  resource.metadata.owner == principal
+};
+
+// allow read if the user is a global admin
+permit(
+  principal in User::"GlobalAdmin",
+  action in [Action::"Read"],
+  resource
+);
+```
+
+Now, we would like to answer the question: given a request that a **User** would like to **Read** a **Document,** what data is required?
+
+To answer the request correctly, Cedar requires two data paths and one entity’s ancestors in the entity store:
+
+```
+Request type: (User, Read, Document)
+Data required:
+   resource.readers
+   resource.metadata.owner
+   principal.**ancestors** (where ancestors are the ancestors in the entity hierarchy)
+```
+
+On the other hand, if the **User** would like to **Edit** a **Document**, we only need one piece of data:
+
+```
+Request type: (User, Edit, Document)
+Data required:
+  resource.metadata.owner
+```
+
+
+
+## Example Usage
+
+![A client communicates with a Cedar Server](clientserver.png)
+
+The image above shows an example usage of the entity manifest. On the left, a client has access to the entity data and a cached entity manifest. On the right, a server stores the schema and policies.
+
+When a client would like to answer a Cedar `Request`, it first consults the cached entity manifest to load the entity slice. Then, it sends the entity slice and the request to the server. The server returns the authorization result.
+
+However, the server or client must take care to ensure that the client’s entity manifest is up-to-date. Since the entity manifest is computed based on the schema and policies, is must be re-computed whenever the schema or policies change. There are multiple ways to ensure that the cache is up to date, one being to tag each entity manifest uniquely and check on each request.
+
+Looking to the future, we hope to provide several options and guidance for how to use the entity manifest:
+
+1. Use the `SimplifiedEntityLoader` api (see later section).
+2. Use a specialized entity loader when the format of the database sufficiently matches the schema.
+3. Write custom entity loading that leverages the entity manifest.
+
+
+
+## Data structure
+
+This document proposes the **Entity Manifest**, a data structure that specifies the data needed for each type of request. To reduce redundant access paths, the manifest is stored as a Trie.
+
+```
+type EntityManifest = HashMap<RequestType, RequestEntityManifest>
+```
+
+`RequestType` stores the principal type, resource type, and action UID. This allows entity slicing to be specific to the type of the request, greatly reducing the amount of data required. (The context type can be inferred from the action UID and schema.)
+
+```
+pub struct RequestType {
+    pub principal: EntityType,
+    pub action: EntityUID,
+    pub resource: EntityType,
+}
```

Comment:
Interesting. Why not allow `EntityUID`s to be passed for the `principal` and `resource`?

Reply:
The `RequestType` is used to index the entity manifest. When it is computed, the actual principal and resource are not known (but the types are). Action ids are known since there are a finite number.